packages/runtime/container-runtime/src/summarizer.ts

```diff
@@ -252,13 +261,34 @@ export class Summarizer extends EventEmitter implements ISummarizer {
         }
     }
 
-    public readonly summarizeOnDemand: ISummarizer["summarizeOnDemand"] = (...args) => {
-        if (this._disposed || this.runningSummarizer === undefined || this.runningSummarizer.disposed) {
-            throw Error("Summarizer is not running or already disposed.");
+    public readonly summarizeOnDemand: ISummarizer["summarizeOnDemand"] = async (...args) => {
+        try {
+            if (this._disposed) {
+                throw Error("Summarizer is already disposed.");
+            }
+            if (this.runningSummarizer?.disposed !== false) {
+                this.runningSummarizer = undefined;
+                const runCoordinator: ICancellableSummarizerController =
+                    await this.runCoordinatorCreateFn(this.runtime);
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                await this.start(this.runtime.clientId!, runCoordinator, { disableHeuristics: true });
```

Comment:
I'm a bit concerned about race conditions here. I.e. can summary manager at the same time start same process and create another instance of running summarizer? I think we should somehow protect against it - either not allow for two to coexists at all, or allow safe creation of a single instance.

Reply:
Is this related to https://github.com/microsoft/FluidFramework/issues/8359? I think one of the concerns there is how a combination of server-controlled (i.e., on demand) and client-controlled (i.e., SummaryManager-controlled, as we have today) summarization would behave, which sounds like what you're talking about. In a totally server-controlled world, where all ContainerRuntime's have disableSummaries=true, there are no SummaryManager's, so what you're describing can't happen. I'm going to have to construct a test case to understand how a combination of the 2 models would behave.

Reply:
This API does is generally exposed, so we can't say that it will be used in some specific scenario. I think the main scenario where it will be used - it tests, where they want to control forcing summaries, while there might be actual summarizer manager / summarizer that also produce summaries on timer / events. So, we need to ensure that such usage is correct.

Reply:
I'm a little unclear about your last comment, but I think we're saying more or less the same thing. Leaving server/client control out of it for a second, on-demand summaries have to be safe if there are SummaryManager's around that want to kick off summaries on their own schedule. It seems like there has to be a central controller that will force a summary to wait if another one is already in progress.

Reply:
The same kind of control would have to apply, it seems, to keep 2 on-demand summaries from happening at the same time. I suppose we could leave it up to the developer who's using on-demand to make sure they do it safely, but if we have to do something anyway to avoid a race condition, it seems like we may as well use the same mechanism to keep 2 on-demand summaries from stepping on each other.

Reply:
I brought this up with @anthony-murphy, and one simple thing we thought we could do in the short term is not allow an on-demand summary if there's an elected summarizer. That would work for the current ask. If we went that route, I would open an issue to craft a friendlier solution after this change. We could also use events. If there's an elected summarizer, the on-demand summarizer notifies it that an on-demand summary is about to begin. At that point, the on-demand summarizer could either charge ahead or wait for the elected summarizer to acknowledge that it received the event. Or, if there's an elected summarizer, summarizeOnDemand() tells the elected summarizer to begin a summary now, but that leaves open the question of what summarizeOnDemand() returns to the caller in that case.